src/hotspot/share/services/heapDumper.cpp

```diff
@@ -1479,6 +1565,138 @@ void HeapObjectDumper::do_object(oop o) {
   }
 }
 
+// The dumper controller for parallel heap dump
+class DumperController : public CHeapObj<mtInternal> {
+ private:
+   Monitor* _lock;
+   const uint   _dumper_number;
+   uint   _complete_number;
+
+ public:
+   DumperController(uint number) :
+     _lock(new (std::nothrow) PaddedMonitor(Mutex::leaf, "DumperController_lock", true, Monitor::_safepoint_check_never)),
+     _dumper_number(number),
+     _complete_number(0) { }
+
+   ~DumperController() { delete _lock; }
+
+   void dumper_complete(DumpWriter* local_writer, DumpWriter* global_writer) {
+     MonitorLockerEx ml(_lock, Mutex::_no_safepoint_check_flag);
+     _complete_number++;
+     // propagate local error to global if any
+     if (local_writer->has_error()) {
+       global_writer->set_error(local_writer->error());
+     }
+     ml.notify();
```

Comment:
notify is not necessary when  _complete_number != _dumper_number

Reply:
Good catch, will ship with optimization patch for usptream.